### DISCUSSION TITLE
Compute tempograms with beat/tempo detection algorithms

### DISCUSSION THREAD

**Discussion by alastair** [CPJKU/madmom#501] [Q&A] [answered]
Hi,
From what I understand from reading some source files in madmom, I can predict downbeats using something like `TCNBeatTracker single song.wav`, and compute a global BPM using `TCNTempoDetector`. It looks like the TempoEstimationProcessor can use any number of TempoHistogramProcessors, which will compute a histogram from these beats, allowing you to pick the highest value as a predicted BPM.

With other BPM algorithms, such as TempoCNN or essentia's RhythmExtractor, I can also get BPM estimations throughout the duration of the song, and plot it. Does madmom have something similar? I considered just taking the time intervals between the detected beats, but I'm not sure if the histogram processors perform any additional smoothing on these intervals before generating the histogram.

Thanks

**Comment by superbock** [comment, marked as answer]
The quickest solution is definitely to compute the inter beat intervals (IBIs) and compute the tempo from there. But this will not yield the same result as running `TempoDetector` or `TCNTempoDetector` on the same song. Sometimes tempo derived from IBIs work quite well, sometimes they don't.

Per default `TempoDetector` uses resonating comb filters to estimate the (local) tempo from the beats, but does so after smoothing both the beat activations and applies some tempo histogram smoothing afterwards. Depending on which `TempoHistogramProcessor` is used (note: with the current source you can only define a single one, not multiples to be used with), `TempoEstimationProcessor` allows processing on `online` mode, i.e. causal processing of the input stream. IIRC, we do not apply any activation smoothing in this mode, so the results are very likely to be less stable. Please note, that if you run `TempoDetector` in lonline mode, it also uses a uni-directional RNN to predict the beats — which yields lower performance compared to the bidirectional (offline) models.

`TCNTempoDetector` on the other hand uses a completely different mechanism: it uses a classification layer (very much like TempoCNN), but does perform this tempo classification on globally aggregated values. You can of course alter the underlying NN to perform this classification for each frame, but lower performance is also to be expected.

So the general answer is: no madmom does not provide a mechanism to yield local tempi, but it is possible to modify the source to do so. However, I expect the results to be worse compared to global tempo estimation.
